### PR TITLE
[GLUTEN-1348][CORE] Fallback if projections in Expand node have scalar function

### PR DESCRIPTION
**Pull request by exmy**
## What changes were proposed in this pull request?

(Fixes: #1348)

## How was this patch tested?

manual test

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala [hunk 1]
```diff
@@ -223,6 +226,13 @@ case class ExpandExecTransformer(projections: Seq[Seq[Expression]],
       return false
     }
 
+    // Fallback if projections have scalar function, which is not supported currently.
+    if (projections.exists(
+          _.exists(expr => !expr.isInstanceOf[Attribute] && !expr.isInstanceOf[Literal]))) {
+      logWarning("There are scalar functions in expand node, which is not supported currently.")
+      return false
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala:233]: @JkSelf Do we have such validation in our native validator for Expand?

**Comment by exmy** [comment]
@JkSelf By the way, I am a little confused about the `ExpandExecTransformer`.
```scala
projections.foreach { projection =>
  val groupExprNodes = new util.ArrayList[ExpressionNode]()
  for (i <- aggSize until (projection.size - 1)) {
    if (!(projection(i).isInstanceOf[Literal] &&
      projection(i).asInstanceOf[Literal].value == null)) {
      var groupExprNode = ExpressionConverter
        .replaceWithExpressionTransformer(
          projection(i),
          originalInputAttributes
        )
```
Here, projections that are not null literals are transformed, which may cause problems in the following case
```shell
ExpandExecTransformer [[uid#12L, dispatch_id#27, null, 1], [uid#12L, null, video_id#23L, 2]]
       ProjectExecTransformer [uid#12L, video_id#23L, dispatch_id#27]
```
in this case, groupsetExprNodes is transformed to something like  `[ [0, 2], [0, 1] ]`, and in the backend it may parsed as `[uid, null, dispatch_id, gid], [uid, video_id, null, gid]`, which is not consistent with original projections. 

Do I understand it right?

**Comment by JkSelf** [comment]
@exmy I don't understand why the `groupsetExprNodes `is transformed to `[ [0, 2], [0, 1] ]`.

This code is to calculate the group cols and filter the gid col. Here the `groupsetExprNodes ` cols is [dispatch_id] and [video_id] for your example. And the child output is `[uid, dispatch_id, video_id]`. Velox backend will judge whether the group key exist in the output. The `groupsets `default value is null and will be set to the group key if exist. Then we will get the value is `[dispatch_id, null]` and `[null, video_id]`.

File: gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala (hunk 1 above)
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala:233]: Looks Velox backend does not need this, is that right?

**Comment by FelixYBW** [comment]
If we solve the fallback, should we add a project before expand? Does Velox's expand support expression?

**Comment by exmy** [comment]
> @exmy I don't understand why the `groupsetExprNodes `is transformed to `[ [0, 2], [0, 1] ]`.
> 
> This code is to calculate the group cols and filter the gid col. Here the `groupsetExprNodes ` cols is [dispatch_id] and [video_id] for your example. And the child output is `[uid, dispatch_id, video_id]`. Velox backend will judge whether the group key exist in the output. The `groupsets `default value is null and will be set to the group key if exist. Then we will get the value is `[dispatch_id, null]` and `[null, video_id]`.

@JkSelf  Sorry if I didn't express myself clearly. What I meant was that `[ [0, 2], [0, 1] ]` is the SelectionNode for the child output references, for example:
```json
"groupings": [{
    "groupSetsExpressions": [{
      "selection": {
        "directReference": {
          "structField": {
          }
        }
      }
    }, {
      "selection": {
        "directReference": {
          "structField": {
            "field": 2
          }
        }
      }
    }]
  }, {
    "groupSetsExpressions": [{
      "selection": {
        "directReference": {
          "structField": {
          }
        }
      }
    }, {
      "selection": {
        "directReference": {
          "structField": {
            "field": 1
          }
        }
      }
    }]
  }]
```
And in this case, the child output is actually `[uid, video_id, dispatch_id]`. The issue is caused by the difference in order between `video_id` and `dispatch_id`, which is also why there may be parsing problems at least in the clickhouse backend. I am not sure if velox also has this issue.

File: gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala (hunk 1 above)
Comment (exmy) [reply gluten-core/src/main/scala/io/glutenproject/execution/ExpandExecTransformer.scala]: > Looks Velox backend does not need this, is that right?

ok, have made the check only in CH backend

**Comment by zhli1142015** [comment]
> > @exmy I don't understand why the `groupsetExprNodes `is transformed to `[ [0, 2], [0, 1] ]`.
> > This code is to calculate the group cols and filter the gid col. Here the `groupsetExprNodes ` cols is [dispatch_id] and [video_id] for your example. And the child output is `[uid, dispatch_id, video_id]`. Velox backend will judge whether the group key exist in the output. The `groupsets `default value is null and will be set to the group key if exist. Then we will get the value is `[dispatch_id, null]` and `[null, video_id]`.
> 
> @JkSelf Sorry if I didn't express myself clearly. What I meant was that `[ [0, 2], [0, 1] ]` is the SelectionNode for the child output references, for example:
> 
> ```json
> "groupings": [{
>     "groupSetsExpressions": [{
>       "selection": {
>         "directReference": {
>           "structField": {
>           }
>         }
>       }
>     }, {
>       "selection": {
>         "directReference": {
>           "structField": {
>             "field": 2
>           }
>         }
>       }
>     }]
>   }, {
>     "groupSetsExpressions": [{
>       "selection": {
>         "directReference": {
>           "structField": {
>           }
>         }
>       }
>     }, {
>       "selection": {
>         "directReference": {
>           "structField": {
>             "field": 1
>           }
>         }
>       }
>     }]
>   }]
> ```
> 
> And in this case, the child output is actually `[uid, video_id, dispatch_id]`. The issue is caused by the difference in order between `video_id` and `dispatch_id`, which is also why there may be parsing problems at least in the clickhouse backend. I am not sure if velox also has this issue.

Hello @exmy, I raised a PR to fix a similar problem for velox, the basic idea is to pass the projections to BE, no need to distinguish the agg cols and group cols. I think this behavior aligns more with spark ExpandExec. I'm verifying it with Velox. Please let me know if this can also help for CH, maybe we can corrdinate the both velox change and CH change.
Thanks.
https://github.com/oap-project/gluten/pull/1361

**Comment by exmy** [comment]
> Hello @exmy, I raised a PR to fix a similar problem for velox, the basic idea is to pass the projections to BE, no need to distinguish the agg cols and group cols. I think this behavior aligns more with spark ExpandExec. I'm verifying it with Velox. Please let me know if this can also help for CH, maybe we can corrdinate the both velox change and CH change. Thanks. #1361

Looks great, it can help for CH. Looking forward to it very much~

**Pull request closed without merging**